### ISSUE TITLE
Snatched books issue

### ISSUE THREAD

**Issue by forge33** [DobyTang/LazyLibrarian#519]
LL seems to find a book, snatch it, queue it up in sabnzbd. However it never marks it as Snatched and doesn't send out notifications of the snatched book. Ends up downloading the book over and over again complaining it hasn't been Snatched.

**Comment by forge33** [comment]
09-Nov-2016 14:27:39 - DEBUG   :: WEBSERVER : Searching for book with id: 26114545
09-Nov-2016 14:27:39 - INFO    :: SEARCHNZB : NZB Searching for 1 book
09-Nov-2016 14:27:39 - DEBUG   :: SEARCHNZB : Using stored capabilities for https://api.nzbgeek.info
09-Nov-2016 14:27:39 - INFO    :: SEARCHTOR : TOR Searching for 1 book
09-Nov-2016 14:27:39 - DEBUG   :: SEARCHNZB : [IterateOverNewzNabSites] - https://api.nzbgeek.info
09-Nov-2016 14:27:39 - DEBUG   :: SEARCHTOR : [IterateOverTorrentSites] - libgen.io
09-Nov-2016 14:27:39 - DEBUG   :: SEARCHNZB : [NewzNabPlus] searchType [book] with Host [https://api.nzbgeek.info] mode [nzb] using api [xxxxx] for item [{'bookSub': u'', 'searchterm': u'Ada Palmer Too Like the Lightning (Terra Ignota, #1)', 'bookName': u'Too Like the Lightning (Terra Ignota, #1)', 'authorName': u'Ada Palmer', 'bookid': u'26114545'}]
09-Nov-2016 14:27:39 - DEBUG   :: SEARCHNZB : [NewzNabPlus] - nzb Search parameters set to {'author': u'Ada Palmer', 'apikey': u'xxxxx', 't': 'book', 'cat': '7000,7020', 'title': u'Too Like the Lightning (Terra Ignota 1)'}
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : Parsing results from <a href="https://api.nzbgeek.info/api?author=Ada+Palmer&apikey=xxxxx&t=book&cat=7000%2C7020&title=Too+Like+the+Lightning+%28Terra+Ignota+1%29">https://api.nzbgeek.info</a>
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : [NewzNabPlus] - result fields from NZB are {'bookid': u'26114545', 'nzbmode': 'nzb', 'nzbdate': 'Mon, 23 May 2016 11:21:28 +0000', 'nzbtitle': 'Ada Palmer - [Terra Ignota 01] - Too Like the Lightning (epub)', 'nzbsize': '1783000', 'nzburl': 'https://api.nzbgeek.info/api?t=get&id=94db297aac39375dbf5d73b3fc615b0f&apikey=xxxxx', 'nzbprov': u'https://api.nzbgeek.info'}
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : Found 1 nzb at https://api.nzbgeek.info for: Ada Palmer Too Like the Lightning (Terra Ignota, #1)
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : NZB author/book Match: 100/97 for Ada Palmer Terra Ignota Too Like the Lightning epub
09-Nov-2016 14:27:41 - INFO    :: SEARCHNZB : Best NZB match (97%): Ada Palmer Terra Ignota Too Like the Lightning epub using book search
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : Ada Palmer Terra Ignota Too Like the Lightning epub adding to wanted
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : Request url for <a href="http://10.42.0.230:8080/api?nzbname=Ada+Palmer+-+Too+Like+the+Lightning+%28Terra+Ignota+%231%29+LL.%2826114545%29&ma_password=xxxxx&apikey=xxxxx&name=https%3A%2F%2Fapi.nzbgeek.info%2Fapi%3Ft%3Dget%26id%3D94db297aac39375dbf5d73b3fc615b0f%26apikey%3Dxxxxx&maxage=1650&cat=books&mode=addurl&output=json&ma_username=forge33">SABnzbd</a>
09-Nov-2016 14:27:41 - DEBUG   :: SEARCHNZB : Result text from SAB: {u'status': True}
09-Nov-2016 14:27:41 - INFO    :: SEARCHNZB : Ada Palmer - Too Like the Lightning (Terra Ignota #1) LL.(26114545) sent to SAB successfully.
09-Nov-2016 14:27:43 - DEBUG   :: SEARCHTOR : Parsing results from <a href="http://libgen.io/search.php?view=simple&open=0&phrase=0&column=def&res=100&req=Ada+Palmer+Too+Like+the+Lightning+%28Terra+Ignota+1%29">libgen</a>
09-Nov-2016 14:27:43 - DEBUG   :: SEARCHTOR : Found 0 results from libgen for Ada Palmer Too Like the Lightning (Terra Ignota 1)
09-Nov-2016 14:27:43 - DEBUG   :: SEARCHTOR : No torrent's found for [Ada Palmer Too Like the Lightning (Terra Ignota 1)] using searchtype book
09-Nov-2016 14:27:43 - DEBUG   :: SEARCHTOR : [IterateOverTorrentSites] - libgen.io
09-Nov-2016 14:27:44 - DEBUG   :: SEARCHTOR : Parsing results from <a href="http://libgen.io/search.php?view=simple&open=0&phrase=0&column=def&res=100&req=Ada+Palmer+Too+Like+the+Lightning">libgen</a>
09-Nov-2016 14:27:44 - DEBUG   :: SEARCHTOR : Found 0 results from libgen for Ada Palmer Too Like the Lightning
09-Nov-2016 14:27:44 - DEBUG   :: SEARCHTOR : No torrent's found for [Ada Palmer Too Like the Lightning] using searchtype shortbook
09-Nov-2016 14:27:44 - DEBUG   :: SEARCHTOR : Searches for Ada Palmer Too Like the Lightning returned no results.
09-Nov-2016 14:27:44 - INFO    :: SEARCHTOR : TORSearch for Wanted items complete, found 0 books

**Comment by philborman** [comment]
Looks like the code talking to sabnzbd is crashing just after it prints the "success" message.
What version of sabnzbd are you using? I'm wondering if not all versions support returning a download id.

**Comment by forge33** [comment]
0.7.20

**Comment by philborman** [comment]
Current is 1.1.0
The documentation says adding a download should return
Result text from SAB: {u'status': True, u'nzo_ids': [u'SABnzbd_nzo_5_hfyr']}
whereas your version returns  
Result text from SAB: {u'status': True}
Note the missing 'nzo_ids' result. This is what we use to identify the download so we can monitor progress. If you can upgrade your sabnzbd the problem will go away, but I will add code to get around the problem in the next version. Alternatively if you don't mind a bit of editing you can modify sabnzbd.py line 112, change return result['nzo_ids'][0]  to  return 'unknown'

**Comment by forge33** [comment]
Thanks
I'll do an upgrade tonight of sab. Maybe throw an error in LL if sab is older than x version?

**Comment by philborman** [comment]
Can't find any documentation saying when the nzb_ids feature was added.
Simplest is just to look if we got an nzb_ids returned, if so use it, if not ignore.

**Comment by forge33** [comment]
https://github.com/sabnzbd/sabnzbd/commit/b8889756d48857e851a6da3925ce57dec45d1e27

looks like 0.8.0 is the earliest it's shown up

**Comment by forge33** [comment]
ok upgrading has worked, now @ 1.1.0

**Comment by philborman** [comment]
Ah, thanks for that. Api documentation doesn't mention it as a new feature
so i assumed it was always there. Lazylibrarian didn't use it until
recently though, it's more useful for torrents than nzb at the moment,  as
we use it to stop seeding after completion, but can also be used as a way
to flag stalled downloads so we can mark them as failed and try another
search.

I've just added  a couple of lines to check if we got an id returned and
set a default value if not. Should get around the problem.

A few other changes need more testing before release though.
